### PR TITLE
storage: de-flake TestStoreRangeRemoveDead

### LINKED ISSUE #30708: storage: TestStoreRangeRemoveDead failed under stress
SHA: https://github.com/cockroachdb/cockroach/commits/9f191cdad64747f7f19ec4990e559cbb19e6d372

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeRemoveDead PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=928632&tab=buildLog

```
I180927 06:07:20.113115 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:21.126234 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:22.138801 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:23.151225 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:24.157603 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:25.165726 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:26.186466 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:27.188758 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:28.191561 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:29.194531 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
W180927 06:07:30.200435 68763 internal/client/txn.go:532  [hb] failure aborting transaction: node unavailable; try another peer; abort caused by: result is ambiguous (server shutdown)
I180927 06:07:30.204310 68763 storage/node_liveness.go:830  [hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (server shutdown)
W180927 06:07:30.220042 68763 internal/client/txn.go:532  [hb] failure aborting transaction: node unavailable; try another peer; abort caused by: node unavailable; try another peer
W180927 06:07:30.220476 68763 storage/node_liveness.go:494  [hb] failed node liveness heartbeat: node unavailable; try another peer
W180927 06:07:30.222988 69459 storage/store.go:1657  [s6,r1/3:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W180927 06:07:30.231595 69479 storage/raft_transport.go:281  unable to accept Raft message from (n4,s4):?: no handler registered for (n1,s1):?
W180927 06:07:30.233204 69373 storage/store.go:3820  [s4] raft error: node 1 claims to not contain store 1 for replica (n1,s1):?: store 1 was not found
W180927 06:07:30.233573 69371 storage/raft_transport.go:583  while processing outgoing Raft queue to node 1: store 1 was not found:
W180927 06:07:30.289119 69506 storage/raft_transport.go:583  while processing outgoing Raft queue to node 5: rpc error: code = Canceled desc = grpc: the client connection is closing:
W180927 06:07:30.294849 69480 storage/raft_transport.go:583  while processing outgoing Raft queue to node 6: rpc error: code = Canceled desc = grpc: the client connection is closing:
--- FAIL: TestStoreRangeRemoveDead (46.88s)
	soon.go:49: condition failed to evaluate within 45s: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
		goroutine 68674 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc420b1e5d0, 0x3bd9f60)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xb5
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3c35e00, 0xc4215163c0, 0xc4219b01e0)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x172
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeRemoveDead(0xc4215163c0)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_raft_test.go:2906 +0x686
		testing.tRunner(0xc4215163c0, 0x34caf48)
			/usr/local/go/src/testing/testing.go:777 +0x16e
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x565

```

### PR DESCRIPTION
**Pull request by tbg**
The test was making a half-hearted attempt at not naming the leaseholder
the dead store, and was failing when those two coincided.

Hard-code the second store as the dead store and actually stop it (so
that the previous failure mode is now just a slower test run). This
also makes the test more realistic since now the dead node is actually
down.

Fixes #30708.

Release note: None

### CODE REVIEW

**Review by petermattis — APPROVED** [review]
:lgtm:

I'm not very familiar with this test, though. @spencerkimball might want to give this a thumbs up as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30713#-:-LNQJ08y5XAfIqR5GapY:beqxja3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

File: pkg/storage/client_raft_test.go [hunk 1]
```diff
@@ -2900,9 +2904,15 @@ func TestStoreRangeRemoveDead(t *testing.T) {
...
 	})
 
-	// Set the dead store and wait for the replica on the dead node to
-	// be removed and the replacement added. Use a store other than the leader.
-	atomic.StoreInt32(&deadStoreID, int32(getRangeMetadata(roachpb.RKeyMin, mtc, t).Replicas[1].StoreID))
+	// Stop a store which will be rebalanced away from. We can't use the very first
+	// one since getRangeMetadata is hard-coded to query that one, so we'll use the
+	// one after that.
+	storeID := mtc.stores[1].StoreID() // 2
+	atomic.StoreInt32(&deadStoreID, int32(storeID))
```
Comment (spencerkimball) [inline pkg/storage/client_raft_test.go]: No need to keep `deadStoreID` anymore. Remove this setting here and remove the variable declaration above.

Comment (spencerkimball) [reply pkg/storage/client_raft_test.go]: You also don't need `storeID`.

**Review by spencerkimball — APPROVED** [review]

File: pkg/storage/client_raft_test.go (hunk 1 above)
Comment (tschottdorf) [reply pkg/storage/client_raft_test.go]: Done, thanks!

**Comment by tschottdorf** [comment]
Also fixed the data race in the test (mtc.stores gets mutated by stopStore).

**Pull request merged**